sql/src/main/java/io/crate/operation/projectors/BlockingSortingQueuedRowDownstream.java

```diff
@@ -0,0 +1,448 @@
+     *
+     * The handle is blocked when the queue has reached MAX_QUEUE_SIZE
+     * and unblocked when the queue has RESUME_AFTER or less rows
+     */
+    public static int MAX_QUEUE_SIZE = 1000;
+    public static int RESUME_AFTER = 10;
+
+    private final Ordering<Object[]> ordering;
+    private final List<BlockingSortingQueuedRowDownstreamHandle> downstreamHandles = new ArrayList<>();
+    private final List<RowUpstream> upstreams = new ArrayList<>();
+    private final AtomicInteger remainingUpstreams = new AtomicInteger(0);
+    private final AtomicBoolean downstreamAborted = new AtomicBoolean(false);
+    private final RowReceiver downstreamRowReceiver;
+    private final SharedArrayRef lowestToEmit;
+    private final int rowSize;
+    private final AtomicInteger runningHandles = new AtomicInteger(0);
+
+    private int pauseCount = 0;
+
+
+    public BlockingSortingQueuedRowDownstream(RowReceiver rowReceiver,
+                                              int rowSize,
+                                              int[] orderBy,
+                                              boolean[] reverseFlags,
+                                              Boolean[] nullsFirst) {
+        this.rowSize = rowSize;
+        List<Comparator<Object[]>> comparators = new ArrayList<>(orderBy.length);
+        for (int i = 0; i < orderBy.length; i++) {
+            comparators.add(OrderingByPosition.arrayOrdering(orderBy[i], reverseFlags[i], nullsFirst[i]));
+        }
+        ordering = Ordering.compound(comparators);
+        lowestToEmit = new SharedArrayRef(rowSize);
+        this.downstreamRowReceiver = rowReceiver;
+        rowReceiver.setUpstream(this);
+    }
+
+    @Override
+    public RowReceiver newRowReceiver() {
+        remainingUpstreams.incrementAndGet();
+        int handleIdx = runningHandles.getAndIncrement();
+        BlockingSortingQueuedRowDownstreamHandle handle = new BlockingSortingQueuedRowDownstreamHandle(
+                handleIdx, this, rowSize);
+        downstreamHandles.add(handle);
+        return handle;
+    }
+
+    public void upstreamFinished() {
+        runningHandles.decrementAndGet();
+        int remaining = remainingUpstreams.decrementAndGet();
+        if (!emit()) {
+            // if downstream is finished, wake every handle up, so it can signal its upstream that it's done
+            for (BlockingSortingQueuedRowDownstreamHandle handle : downstreamHandles) {
+                if (!handle.isFinished() && handle.paused) {
+                    handle.resume();
+                }
+            }
+        }
+        if (remaining == 0) {
+            LOGGER.trace("paused {} times, per handle rate: {}", pauseCount, pauseCount/(double)downstreamHandles.size());
+            downstreamRowReceiver.finish();
+        }
+    }
+
+    public void upstreamFailed(Throwable throwable) {
+        runningHandles.decrementAndGet();
+        downstreamAborted.compareAndSet(false, true);
```

Comment:
either you can use downstreamAborted.set(true) or you'd have to check the return value and handle the case where downstreamAborted was already true